src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -7164,12 +7164,54 @@ bool canFollowNullableType()
-                        // `?:` operator if possible.
-                        //
-                        // Similarly, if we have `T?[]` we do want to treat that as an array of nullables (following
+                        // These contexts might be a type that is at the end of an expression.
+                        // For backward compatibility we want to consider a `?` token as part of the `?:` operator if possible.
+                        // However, if current token is an identifier, it might be beneficial to allow `?`
+                        // and treat this identifier as a designation. Nullable types in patterns are semantically invalid,
+                        // so we will get a nice error about that during binding
+                        if (this.CurrentToken.Kind == SyntaxKind.IdentifierToken)
+                        {
+                            // Given that we are deciding whether we take `?:` operator or not while looking at state after `?`,
+                            // our focus is on `:` token as well. But `:` have special meaning in context of switch statements
+                            // and expressions (since `:` in switch expressions is recovered to `=>`). So consider the following case:
+                            // `case T? t:`. On its own `T ? t :` can be treated as a conditional expression with condition `T`,
+                            // 'whenTrue' `t` and 'whenFalse' not yet parsed. But it is way better to take `:` as case label end and
+                            // thus parse label condition as a declaration pattern `T? t`. `whenIsKeyword` flag can conveniently tell
+                            // whether we are in a 'top-level' state of a label, so if condition get parenthesized like `case (a ? b : c):`
+                            // we don't take the first `:` as label end and parse label condition as a parenthesized conditional expression
+                            if ((_termState.HasFlag(TerminatorState.IsExpressionOrPatternInCaseLabelOfSwitchStatement) ||
+                                 _termState.HasFlag(TerminatorState.IsPatternInSwitchExpressionArm)) && whenIsKeyword)
+                            {
+                                return true;
+                            }
+
+                            // Special case for `await`: if current token is an `await` identifier and the next token
+                            // can start an expression, prefer parsing conditional expression
+                            // rather than having `await` as a pattern designation
+                            if (this.CurrentToken.ContextualKind == SyntaxKind.AwaitKeyword)
+                            {
+                                using var _ = GetDisposableResetPoint(resetOnDispose: true);
+                                this.EatToken();
+                                return !CanStartExpression();
```

Comment:
honestly, don't even bother with CanStartExpression.  just literally jsut always treat `await` as the start of an expr.  No one is realistically writing `if (x is Foo? await)`.  And, if they do, it's fine to not have that be thought of as a designation.

Reply:
Uh, I added these obscure test cases right before you left that review. `CanStartExpression` in most cases looks only 1 token ahead and in others just several ones, so this check is cheap. Can we leave it as is, so I don't have to redo that previously added tests?